operator/cluster/cluster.go

```diff
@@ -202,6 +189,49 @@ func AddClusterBase(clientset *kubernetes.Clientset, client *rest.RESTClient, cl
 
 }
 
+//getOrCreatePVC will first verify whether the pgcluster name matches the "current primary" CRD value of the pgcluster CRD.
+//Based on this information, the underlying PVC name for the primary pod can be validated. If the expected PVC exists,
+//the correct PVC name is returned.
+//If the "current primary" label and the pgcluster name do match, this is either the recreation of a cluster that has not
+//previously failed over, or it is a new pgcluster. If a recreation, the PVC is reused, otherwise a new PVC is allocated.
+func getOrCreatePVC(clientset *kubernetes.Clientset, cl *crv1.Pgcluster, namespace string) string {
+	//if the current primary label does not equal the base pgcluster name and it is not empty,
+	//then the pgcluster has failed over at some point
+	if cl.Spec.CurrentPrimary != cl.Spec.Name && cl.Spec.CurrentPrimary != "" {
+		//ensure the expected PVC exists
+		_, found, err := kubeapi.GetPVC(clientset, cl.Spec.CurrentPrimary, namespace)
+		if err != nil {
+			log.Errorf("error while trying to find the PVC %s during creation of cluster %s. Error: %s", cl.Spec.CurrentPrimary, cl.Spec.Name, err)
+			publishClusterCreateFailure(cl, err.Error())
+			return ""
+		}
+		if found {
+			log.Debugf("pvc [%s] already present from previous cluster with this same name, will not recreate", cl.Spec.CurrentPrimary)
+			return cl.Spec.CurrentPrimary
+		}
+	}
+
+	//otherwise, if the current primary name matches the PVC name, this cluster is either new, or has not previously had a failover
+	//also, since the PVC not being found is returned as an error and the relevant errors are logged in GetPVC, error
+	//information will not be captured here
+	_, found, _ := kubeapi.GetPVC(clientset, cl.Spec.Name, namespace)
```

Comment:
Should the error be captured and handled here, i.e. in case something goes wrong _other_ than not being able to find the PVC?

Reply:
yeah, I was thinking we might want to continue on and try to create the PVC, but it does make more sense to handle the error and return since we can't be sure of whether the PVC is there or not if GetPVC had an error. I added an error catch similar to the one in the previous block:
```
if err != nil {
		log.Errorf("error while trying to find the PVC %s during creation of cluster %s. Error: %s", cl.Spec.Name, cl.Spec.Name, err)
		publishClusterCreateFailure(cl, err.Error())
		return ""
	}
```
Does this work for you?

Reply:
The PVC check/creation process has been completely overhauled as described in other comments. Resolving this one as it is OBE.